src/CanvasRender/index.js

```diff
@@ -8,6 +8,23 @@ import {WIDTH, HEIGHT} from '../chartCore/defaults'
 
 import basicRender from './basicRender'
 
+let devicePixelRatio
+let backingStoreRatio
+function getScaleRatio() {
+  if (devicePixelRatio && backingStoreRatio) return devicePixelRatio / backingStoreRatio
+  devicePixelRatio = typeof window === 'object' ? window.devicePixelRatio : 1
+  const ctx = typeof document === 'object' ? document.createElement('canvas').getContext('2d') : {}
+  backingStoreRatio =
+    ctx.webkitBackingStorePixelRatio ||
+    ctx.mozBackingStorePixelRatio ||
+    ctx.msBackingStorePixelRatio ||
+    ctx.oBackingStorePixelRatio ||
+    ctx.backingStorePixelRatio ||
+    1
+
+  return devicePixelRatio / backingStoreRatio
+}
+
```

Comment:
Looks great! How about we move this to another file -- maybe `scaleRatio.js` -- which just computes and exports the scale ratio as a constant. That way we don't have the additional function logic.

Reply:
Yup, that sounds good. Do we want to prefer localized utils here or does it make sense to have it in the shared `utils` folder if it's only going to be used here?

Reply:
Let's start out with it localized. I'd like to do a big directory/file reorg at some point.